### DISCUSSION TITLE
Unit  Hook

### DISCUSSION THREAD

**Discussion by thjkhdf12** [Phobos-developers/Phobos#500] [Q&A] [answered]
I tried to add convert warhead, but I don't know how to create hooks of UnitClass and InfantryClass , I want to add a function to apply convert  warhead.

**Comment by Metadorius** [comment, marked as answer]
Is this what are you after? https://github.com/Phobos-developers/Phobos/pull/369

**Reply by thjkhdf12** [reply]
This  usually cause fatal error(Like G.I. convert to Chrono Legionnaire)

**Reply by Metadorius** [reply]
You gotta debug the fatal error then. You can load the crash dump generated with Ares or attach a debugger to a live instance if you compile it with a debug config.

**Reply by MortonPL** [reply]
Changing the Locomotor via warhead is very risky, as it can happen while the target is on the move. See [Ares type converison comments on that](https://ares-developers.github.io/Ares-docs/new/typeconversion.html). #369 already has everything required implemented to function at all (or at least I hope so), so crashes may be a result of changing "illegal" ini entries, though that's very much unexplored territory.
If you want to improve something, use that PR instead  of creating a separate one.

**Reply by thjkhdf12** [reply]
Thanks. I also found it's risky, I modified C++ code and tried to get safe convert warhead but always cause fatal error or keep original locomotor like your convert warhead. Libra can't change locomotor to teleporter or G.I. convert to Chrono Legionnaire will cause fatal error while attacking. So I tried to recreate unit on same location. But it seems useless because exist function AnimationCreateUnit, covert should keep some information (like related trigger).

**Reply by thjkhdf12** [reply]
I am not good at English...

**Reply by Otamaa** [reply]
> Thanks. I also found it's risky, I modified C++ code and tried to get safe convert warhead but always cause fatal error or keep original locomotor like your convert warhead. Libra can't change locomotor to teleporter or G.I. convert to Chrono Legionnaire will cause fatal error while attacking. So I tried to recreate unit on same location. But it seems useless because exist function AnimationCreateUnit, covert should keep some information (like related trigger).

changing FootClass locomotor require step and right procedure , also not all locomotor is swappable afaik , been experimenting on these for a  while now , if you want to know how to sawap locomotor , here is ares code for reference :
https://github.com/Ares-Developers/Ares/blob/f642263ec12dd393bdc02ba122e63d87cd09faaa/src/Ext/Techno/Body.cpp#L524 

This one significantly old code , but can be adapted like droppod above : 
https://github.com/Ares-Developers/Ares/blob/8727708db66760b9eb3e55e22d66a6af9792bdac/src/Ext/Infantry/Hooks.Jumpjet.cpp#L342

Hope these data can help you !

**Reply by thjkhdf12** [reply]
> > Thanks. I also found it's risky, I modified C++ code and tried to get safe convert warhead but always cause fatal error or keep original locomotor like your convert warhead. Libra can't change locomotor to teleporter or G.I. convert to Chrono Legionnaire will cause fatal error while attacking. So I tried to recreate unit on same location. But it seems useless because exist function AnimationCreateUnit, covert should keep some information (like related trigger).
> 
> changing FootClass locomotor require step and right procedure , also not all locomotor is swappable afaik , been experimenting on these for a while now , if you want to know how to sawap locomotor , here is ares code for reference : https://github.com/Ares-Developers/Ares/blob/f642263ec12dd393bdc02ba122e63d87cd09faaa/src/Ext/Techno/Body.cpp#L524
> 
> This one significantly old code , but can be adapted like droppod above : https://github.com/Ares-Developers/Ares/blob/8727708db66760b9eb3e55e22d66a6af9792bdac/src/Ext/Infantry/Hooks.Jumpjet.cpp#L342
> 
> Hope these data can help you !

These are very helpful to me! Thanks for your help!

**Reply by thjkhdf12** [reply]
I used function ChangeLocomotorTo() but useless, maybe some bugs in my code... I'll check it out.
